check_copyright.py

```diff
@@ -281,9 +283,11 @@ def has_valid_copyright(file_name: str, mime: str, is_on_ignore: bool, config_se
                 if mime == MIME['python']:
                     template = '# SPDX-FileCopyrightText: ' + config_section['espressif_copyright']
                 candidate_line = template.format(years=format_years(years[0], file_name))
-                no_time_update = template.format(years=format_years(years[0], file_name, years[1]))
-                if code_lines[comment.line_number() - 1] != no_time_update:
-                    # update the line only in cases when not only the dates are changing
+                no_time_update = template.format(years=format_years(years[0], file_name, years[1] or years[0]))
+                if code_lines[comment.line_number() - 1] != no_time_update or lines_changed >= args.lines_changed:
+                    # update the line only in cases when not only the dates are changing or
```

Comment:
@fhrbata Could you please elaborate on the first rule? 

Am I correct that we update the line only if either the dates are changing or the number of changed lines is greater than or equal to the limit specified by `args.lines_changed`?

Won't the first rule always be `true`?

Reply:
@kumekay The logic, as explained to me by @dobairoland , is that we don't want to automatically update the copyright notice if only the year(s) would change, unless there is some bigger change in the code.  

The current implementation, without this change, will not update the years, unless there is some other change in the copyright notice. Let's say that the current copyright in a file is
`SPDX-FileCopyrightText: 2010-2020 Espressif Systems (Shanghai) CO LTD`
and the script would change it only to
`SPDX-FileCopyrightText: 2010-2023 Espressif Systems (Shanghai) CO LTD`
In this case the years will not be updated, because the `no_time_update` will be the same as the copyright notice in the file. Note that this wasn't working the same way for single year, but it should be fixed with this PR too.
OTH  e.g. for
`SPDX-FileCopyrightText: 2010-2020 Espressif Systems`
 the script would change it to
`SPDX-FileCopyrightText: 2010-2023 Espressif Systems (Shanghai) CO LTD`
because of the missing `(Shanghai) CO LTD`

My understanding is that people were complaining that the script is changing the years in copyright even if they did only some trivial change. Hence this `no_time_update` check was introduced. But we probably want to update the years if there is some bigger change to the file. And that's what this PR should help with.

IIUC by the first rule you mean `code_lines[comment.line_number() - 1] != no_time_update`. The `years` used for `no_time_update` are taken from the original copyright line, so if there is no other change in the copyright compared to the template, they should be the same(`false`) and the second check for changed lines should kick in and enforce the years update if necessary.

Many thanks, hopefully I haven't missing something.

Reply:
@fhrbata Thank you so much for the explanation, LGTM!